extlib/millhone/src/cmd/analyze_jars.rs

```diff
@@ -0,0 +1,188 @@
+    // Visit each layer and fingerprint the JARs within.
+    info!("inspecting container");
+    // May have to make two copies of this Archive, idk if it can be iterated twice.
+    let layers = list_container_layers(image_path)?;
+    let mut discoveries = HashMap::new();
+
+    let mut image = unpack(image_path)?;
+    for entry in image.entries().context("iterate entries")? {
+        let entry = entry.context("read entry")?;
+        let path = entry.path().context("read path")?;
+        if !layers.contains(path.as_ref()) {
+            debug!(?path, "skipped: not a layer file");
+            continue;
+        }
+
+        let layer = path.to_path_buf();
+        // Layers should have a form like blob
+        let layer_discoveries =
+            jars_in_layer(entry).with_context(|| format!("read layer '{layer:?}'"))?;
+        discoveries.insert(LayerPath(layer), layer_discoveries);
+    }
+
+    Ok(discoveries)
+}
+
+/// Open and unpack a file and put it into a tar.
+/// This is done repeatedly because `entries()` can only be read once from an Archive.
+#[tracing::instrument]
+fn unpack(path: &PathBuf) -> Result<Archive<File>> {
+    let file = File::open(path).context("open tar file")?;
+    Ok(tar::Archive::new(file))
+}
+
+#[tracing::instrument(skip(entry))]
+fn jars_in_layer(entry: Entry<'_, impl Read>) -> Result<Vec<DiscoveredJar>> {
+    let mut discoveries = Vec::new();
+
+    let mut entry_archive = tar::Archive::new(entry);
+    for entry in entry_archive.entries().context("list entries in layer")? {
+        let entry = entry.context("read entry")?;
+        let path = entry.path().context("read path")?;
+        if !path.to_string_lossy().ends_with(".jar") {
+            debug!(?path, "skipped: not a jar file");
+            continue;
+        }
+
+        let path = path.to_path_buf();
+
+        info_span!("jar", ?path).in_scope(|| -> Result<()> {
+            debug!("fingerprinting");
+            let entry = buffer(entry).context("Read jar file")?;
+
+            match Combined::from_buffer(entry) {
+                Ok(fingerprints) => discoveries.push(DiscoveredJar::new(path, fingerprints)),
+                Err(e) => warn!("failed to fingerprint: {e:?}"),
+            }
+
+            Ok(())
+        })?;
+    }
+
+    Ok(discoveries)
+}
+
+#[tracing::instrument]
+fn list_container_layers(layer_path: &PathBuf) -> Result<HashSet<PathBuf>> {
+    let mut layers = HashSet::new();
+
+    let mut container = unpack(layer_path)?;
+    for entry in container.entries().context("list entries")? {
+        let entry = match entry {
+            Ok(entry) => entry,
+            Err(e) => {
+                warn!("failed to read entry: {e:?}");
+                continue;
+            }
+        };
+
+        let path = match entry.path() {
+            Ok(path) => path,
+            Err(e) => {
+                warn!("Entry path {:?}", e);
```

Comment:
[nit] inline args to format macros; also warnings _are not_ reported with error report mechanisms so should contain more explanation:
```suggestion
                warn!("Failed to read entry path: {e:?}");
```